### DISCUSSION TITLE
gitStream Feature Announcements

### DISCUSSION THREAD

**Discussion by BenLloydPearson** [linear-b/gitstream#102] [Announcements]
Thanks for your interest in gitStream! If you want to receive updates whenever we add new features, subscribe to this thread to get notifications.

**Comment by vim-zz** [comment]
# ⚡️ New: triggering checks on PR title, descriptions, or labels changes

By default, gitStream evaluates any new commit that is pushed to the PR, triggering automation evaluation.

Since last update, if any of the automation rules reference the [pr context](https://docs.gitstream.cm/context-variables/), gitStream shall trigger and will initiate automation rules evaluation even where there are changes to the PR title, descriptions, or labels.

This allows for greater flexibility in the automation process, ensuring that the relevant automation rules are evaluated and triggered when necessary. The execution model ensures that the automation process is streamlined, efficient, and effective.

**Comment by vim-zz** [comment]
# 🧷 Better: fail safe on gitStream checks 

gitStream checks now have a 10-minute timeout for fail-safe reasons. If the check exceeds this time limit, the result will be displayed as Neutral - Skipped.

**Comment by vim-zz** [comment]
# 🌍 Easy to scale with org level rules

Global rules are now available! Organization rules are ideal when you want to enforce consistent rules across every repo in your organization. To define organization-wide rules, create a special repository named `cm` in your GitHub organization where you can add automation files that will apply to all repositories within that organization. 

[Get started by choosing GitHub Organization Setup](https://docs.gitstream.cm/github-installation/)

**Comment by BenLloydPearson** [comment]
# 🐍 Python support for isFormattingChange filter function 

The [isFormattingChange filter function](https://docs.gitstream.cm/filter-functions/#isformattingchange) now supports the detection of non-functional changes to Python code. You can find an example of how to use this in the [automation library](https://docs.gitstream.cm/automations/approve-python-formatting-change/)

**Comment by vim-zz** [comment]
# 💬 PR context got comments, conversations and reviews

`pr.comments` includes all the PR comments in a list of [Comment structutre](https://docs.gitstream.cm/context-variables/#comment-structure)
`pr.conversations ` includes all the PR conversations in a list of [Conversation structutre](https://docs.gitstream.cm/context-variables/#conversation-structure)
`pr.reviews ` includes all the PR reviews in a list of [Review structutre](https://docs.gitstream.cm/context-variables/#review-structure)

**Reply by AnkurAgarwal-ST** [reply]
Will it be able to pick the comment not yet been submitted and is still in a pending state?

**Reply by AnkurAgarwal-ST** [reply]
Do created_at and updated_at reflect the original time when the comment was made or when the reviews were submitted?

E.g. i gave 1 comment yesterday(15 may) in a pr  and submitted them today(16 may), which date will appear in created_at ?

![image](https://github.com/linear-b/gitstream/assets/89902473/e4fdc621-05b5-4e33-a172-d542553cabe5)

**Reply by vim-zz** [reply]
> Do created_at and updated_at reflect the original time when the comment was made or when the reviews were submitted?

when the comment was made

**Reply by vim-zz** [reply]
> Will it be able to pick the comment not yet been submitted and is still in a pending state?

You wont get the un-submitted comments, but you will be able to see that `pr.reviews` holds a `Review` with pending state. For example assuming the first reviewer has unsubmitted comments while reviewing the PR, then the PR context will look like this `pr.reviews[0].state == 'pending'`

**Reply by AnkurAgarwal-ST** [reply]
Thanks , it helps. Will it trigger a gitstream rules execution as soon as a pending comment is added?

**Reply by vim-zz** [reply]
> Will it trigger a gitstream rules execution as soon as a pending comment is added?

When your CM file has an expression that includes `pr.reviewes` gitStream will be triggered for any PR review, including pending reviews

**Reply by AnkurAgarwal-ST** [reply]
Thanks a lot for the clarification @vim-zz

**Reply by vim-zz** [reply]
@AnkurAgarwal-ST after some more tests, GitHub is not sending an event for pending reviews 😞

**Reply by AnkurAgarwal-ST** [reply]
CR for github 😅 ,

 But I assume pr. reviews will still be able to give the pending comment. how should the trigger be done needs to be sorted since we don't have events from github.

**Comment by BenLloydPearson** [comment]
Hey there, /:\ gitStream folks! :wave:

We've been cooking up a storm behind the scenes, and we realize we've been a bit quiet about some of our recent updates—let's fix that!

- **New [Integrations](https://docs.gitstream.cm/integrations/)** :rocket: - We've paved the way for connections with third-party tools with brand new integrations with [SonarCloud](https://docs.gitstream.cm/integrations/sonar/), [Dependabot](https://docs.gitstream.cm/integrations/dependabot/), [Jit](https://docs.gitstream.cm/integrations/jit), and [Swimm](https://docs.gitstream.cm/integrations/swimm/). This is just the beginning, expect more developer tools to join our gitStream ensemble!

- **[Automation Library](https://docs.gitstream.cm/automations/automation-library/)** :books: - These examples are for those seeking to maximize their gitStream experience or seeking sparks of inspiration for their next automation. Have an innovative concept for our library? Feel free to [submit it](https://github.com/linear-b/gitstream/issues/new?assignees=&labels=new-example&projects=&template=new_automation_example.md&title=New+Example%3A+) and will give you a shoutout for your help!

- **gitStream UI** :computer: - Curious about which gitStream automations are getting this most use, or the amount of time gitStream has freed up for you? Unleash these insights with the sleek, new gitStream UI! Visit https://app.gitstream.cm/ to get started. For more info, check out the [details here](https://docs.gitstream.cm/how-it-works/#gitstream-ui).

- **[GitLab](https://docs.gitstream.cm/gitlab-installation/)** :tada: - Drumrolls for the latest addition to our supported platforms!

New Automation examples on the block:
* **[Colors custom expression](https://docs.gitstream.cm/automations/utilities/colors-custom-expression/)** - Effortlessly create labels that blend perfectly with GitHub's default color palette. This can be applied to any existing automation you have that uses labels.
* **[CM file header](https://docs.gitstream.cm/automations/utilities/cm-header/)** - Add a descriptive header to your CM files to make your future self happy when you need to update your automations.

This summer, gear up to discover new avenues to customize gitStream to your needs, incorporate a wider range of third-party developer tools, and establish more dynamic links to your CI pipelines.

**Comment by BenLloydPearson** [comment]
# New Automation Action ⚙️  Explain Code Experts

Users reported that gitStream can be noisy and difficult to follow when updating PR comments that leverage the `explainCodeExperts` filter function. To solve this, we've released `explain-code-experts@v1` as an automation action that will post a single comment to relevant PRs and update the existing message to match further updates to the PR. In the past, gitStream would create a new comment and delete the old comment to ensure consistency, but this new automation action eliminates the need for gitStream to do this.

We recommend upgrading to this if you have any automations that leverage the `explainCodeExperts` filter function to post comments.

Here's an example of how `explain-code-experts` works:

```
automations:
  code_experts:
    if:
      - true
    run:
      - action: explain-code-experts@v1
        args:
          gt: 10
```

**Comment by BenLloydPearson** [comment]
# Plugins, Integrations, Outgoing Webhooks, and More

It’s been a while since we updated our community on the wealth of new features we have in gitStream, so I want to take a moment to cover all the new things we’ve been working on with gitStream over the last six months. 

We want gitStream to be the ultimate workflow automation tool for your git repository. To that end, we’ve been focusing on improving overall extensibility to make it easier to connect gitStream to external services.

The most significant new feature in the last six months has been [Filter Function Plugins](https://docs.gitstream.cm/plugins/), which enable you to use JavaScript plugins to extend functionality for more advanced data processing and pulling data from external APIs. Coupled with the ability to send outgoing webhooks, you now have greater control over the workflow automations you can implement for your git repos. Soon, we’ll also open up plugin capabilities to automation actions, completing our vision of giving you complete automation control for your PR processes.

Here are some other significant new features:

* [Playground](https://docs.gitstream.cm/gitStream-playground/) - Quickly build new automations and test on real-world PRs with instant syntax error highlighting.
* [Dashboard](https://app.gitstream.cm/) - Get insights on how many gitStream automations your organization has running over time.
* New automation actions: 
    * [send-http-request](https://docs.gitstream.cm/automation-actions/#send-http-request) - Send outgoing webhooks from your GitHub repo to external APIs and services.
    * [run-github-workflow](https://docs.gitstream.cm/automation-actions/#run-github-workflow) - Manually dispatch GitHub actions to selectively trigger CI workflows based on the contents of a PR.
    * [send-slack-message](https://docs.gitstream.cm/automation-actions/#send-slack-message) - Send slack notifications that includes helpful metadata from your git repo

Additionally, we have a wealth of new content on the [gitStream integrations](https://docs.gitstream.cm/integrations/) page that demonstrates workflow automations for the following use cases:

* Use gitStream to improve engineering productivity metrics tracking in [LinearB](https://docs.gitstream.cm/integrations/linearb). 
* Use gitStream to trigger [GitHub Actions](https://docs.gitstream.cm/integrations/github-actions) workflows for more efficient CI orchestration selectively.
* Trigger workflow automations based on the results of [Orca Security](https://docs.gitstream.cm/integrations/orca-security) scans.
* Automatically merge PRS from [Dependabot](https://docs.gitstream.cm/integrations/dependabot), [Renovate](https://docs.gitstream.cm/integrations/renovate), and [Snyk](https://docs.gitstream.cm/integrations/snyk)
* Automatically label PRs that contain [Copilot-assisted code](https://docs.gitstream.cm/integrations/github-copilot).
* Here are new examples of automatically linking PRs to [Asana](https://docs.gitstream.cm/integrations/asana), [Shortcut](https://docs.gitstream.cm/integrations/shortcut), and [Azure Boards](https://docs.gitstream.cm/integrations/azure-boards) resources.
* Send notifications from your git repos to [Slack](https://docs.gitstream.cm/integrations/slack) and [Microsoft Teams](https://docs.gitstream.cm/integrations/teams).
* Flag PRs that contain feature flags from [DevCycle](https://docs.gitstream.cm/integrations/devcycle),[ Launch Darkly](https://docs.gitstream.cm/integrations/launch-darkly), [Flagsmith](https://docs.gitstream.cm/integrations/flagsmith), [Unleash](https://docs.gitstream.cm/integrations/unleash), and [Flagr](https://docs.gitstream.cm/integrations/flagr).
* Send PR data to [Zapier](https://docs.gitstream.cm/integrations/zapier) to connect your git repo to thousands of external tools.

We’re hard at work building some extremely exciting improvements to the gitStream ecosystem. Stay tuned for further updates!

**Comment by BenLloydPearson** [comment]
# New Trigger Conditions, GitLab On-Prem Beta, and Jira Examples

Here's the latest updates from the gitStream team at LinearB:

* gitStream now supports the option to trigger automations when PRs are ready for review, closed, or reopened, providing additional flexibility for a wide range of use cases. [Read Execution Model](https://docs.gitstream.cm/execution-model/) for more details.
* New triggers let you [specify which branches and repos automations should run on](https://docs.gitstream.cm/execution-model/#triggers), providing greater control over how you customize automations.
* We published a handful of new examples to the [Jira integration page](https://docs.gitstream.cm/integrations/jira/) that show how to automatically create and update Jira tickets based on PR status and more.
* GitLab Self-Hosted: Beta - gitStream now supports GitLab on-prem! This is a beta release and you can find details on the [GitLab installation page](https://docs.gitstream.cm/gitlab-installation/). Please report any issues to the gitStream GitHub repo.
* [Cache gitStream Docker Image for faster performance](https://docs.gitstream.cm/github-installation/#caching-the-docker-image-using-github-cache) - You can now configure the Github Action workflow file for gitStream to use a cached version of the Docker container. This is useful for gitStream installations with high volumes of PRs where GitHub Action efficiency is a concern.

**Comment by BenLloydPearson** [comment]
# gitStream GitHub App v2

We've released version 2 of the gitStream app for GitHub. This update brings improved performance, a simplified configuration, and enhanced compatibility options. This update also enables you to run gitStream on self-hosted GitHub Actions runners.

[You can find more details about this release here](https://github.com/linear-b/gitstream-github-action/releases/tag/v2).

**Comment by BenLloydPearson** [comment]
# Upcoming Change to gitStream Free Plan: 250 PR Limit per Month

At LinearB, we’re committed to making gitStream an essential automation tool for engineering teams. As gitStream adoption has grown, we’ve seen incredible use cases that streamline PR workflows and improve developer efficiency.

To ensure we can continue providing a high-quality experience for all users, we’re introducing a **250 PR per month limit** for free-tier gitStream users, effective **February 3rd, 2025**. This change allows us to maintain performance and fairness while supporting the continued growth of our automation platform.

**What this means for free users:**



* You can still automate up to **250 PRs per month** at no cost.
* Your teams will receive **usage notifications at both the 90% and 100% thresholds**.
* If you exceed this limit, gitStream automation will pause until the beginning of the next month.
* Teams needing higher limits can [contact LinearB](https://linearb.io/contact-us) to discuss paid plans for expanded usage.

We understand that changes like these can impact workflows, and we’re here to help. If you have any questions or need guidance on managing your PR volume, **feel free to reach out to us at product@linearb.io.**

Thank you for being part of the gitStream community, we’re excited to continue improving developer workflow automation together!

**Reply by pixochi** [reply]
Hi, just wanted to let you know that the email address `product@linearb.io` is unreachable, sending an email to the address fails with `Delivery Status Notification (Failure)`. Is there a new contact we can reach out to?
